src/Bicep.Core.IntegrationTests/ScenarioTests.cs

```diff
@@ -3040,5 +3040,128 @@ public void Test_Issue_6224()
             var compilation = new Compilation(features, BicepTestConstants.NamespaceProvider, sourceFileGrouping, configuration, BicepTestConstants.LinterAnalyzer);
             compilation.GetEntrypointSemanticModel().GetAllDiagnostics().Should().NotBeEmpty();
         }
+
+        /// <summary>
+        /// https://github.com/Azure/bicep/issues/3169
+        /// </summary>
+        [TestMethod]
+        public void Test_Issue_3169()
+        {
+            var result = CompilationHelper.Compile(@"
+resource newStg 'Microsoft.Storage/storageAccounts@2021-04-01' = {
+  name: 'test'
+  kind: 'StorageV2'
+  location: resourceGroup().location
+  sku: {
+    name: 'Standard_LRS'
+  }
+}
+
+resource existingStg 'Microsoft.Storage/storageAccounts@2021-04-01' existing = {
+  name: newStg.properties.accessTier
+}
+");
+            result.ExcludingLinterDiagnostics().Should().NotHaveAnyDiagnostics();
+        }
+
+        /// <summary>
+        /// https://github.com/Azure/bicep/issues/3169
+        /// </summary>
+        [TestMethod]
+        public void Test_Issue_3169_2()
+        {
+            var result = CompilationHelper.Compile(@"
+resource newStg 'Microsoft.Storage/storageAccounts@2021-04-01' = {
+  name: 'test'
+  kind: 'StorageV2'
+  location: resourceGroup().location
+  sku: {
+    name: 'Standard_LRS'
+  }
+}
+
+resource existingStg 'Microsoft.Storage/storageAccounts@2021-04-01' existing = {
+  name: newStg.properties.accessTier
+}
+
+resource foo 'Microsoft.Storage/storageAccounts@2021-04-01' = {
+  name: existingStg.name
+  kind: 'StorageV2'
+  location: resourceGroup().location
+  sku: {
+    name: 'Standard_LRS'
+  }
+}
+");
+            result.ExcludingLinterDiagnostics().Should().HaveDiagnostics(new[] {
+                ("BCP120", DiagnosticLevel.Error, "This expression is being used in an assignment to the \"name\" property of the \"Microsoft.Storage/storageAccounts\" type, which requires a value that can be calculated at the start of the deployment. Properties of existingStg which can be calculated at the start include \"apiVersion\", \"id\", \"type\".")
+            });
+        }
```

Comment:
Can we add another test case where a new resource's name references an existing resource's name that is a DTC with following assertions?
- The result contains no diagnostics
- In the emitted JSON template, the new resource's name is correctly inlined (not a reference function).